README.md

```diff
@@ -9,16 +9,100 @@ Versions 1+ are hybrid elements that support Polymer 1 and Polymer 2.
 - [Demo](https://www.webcomponents.org/element/jstnhuang/ros-rviz/demo/demo/index.html).
   Note that the demo requires a secure websocket server, since it is served over HTTPS.
 
-To run the demo, run rosbridge, tf2_web_republisher, and the interactive marker proxy as needed.
+To run the demo, run `rosbridge`, `tf2_web_republisher`, and the interactive marker proxy as needed.
 
-Example:
-```
-roslaunch pr2_gazebo pr2_empty_world.launch
-roslaunch rosbridge_server rosbridge_websocket.launch
-rosrun tf2_web_republisher tf2_web_republisher
-```
+## Quick start
+
+1. Install dependencies:
+
+        curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.4/install.sh | bash
+        . ~/.bashrc
+        nvm install node
+        npm install -g bower polymer-cli
+
+1. Clone this repo:
+
+        git clone https://github.com/osrf/polymer-ros-rviz
+        cd polymer-ros-rviz
+
+1. Install:
+
+        bower install
+
+1. Serve demos and docs:
+
+        polymer serve --hostname 0.0.0.0
+
+   > The hostname is recommended if you're running from inside docker.
+
+   > By default, it will serve on port 8081
+
+1. View demo:
+
+    http://localhost:8081/components/ros-rviz/demo/index-no-data-binding.html
+
+    > The `no-data-binding` demo is recommended since it works on Firefox, see issue #1
+
+So far, there's nothing running in the backend, so the websocket has nothing to
+connect to.
+
+Let's run an example:
+
+1. Install dependencies (assumes you already have core ROS Kinetic packages installed):
+
+        sudo apt install -y \
+          ros-kinetic-pr2-gazebo \
+          ros-kinetic-rosbridge-server \
+          ros-kinetic-tf2-web-republisher
+
+1. Install the Caddy web server, which will help us serve PR2 meshes with appropriate CORS settings:
+
+        curl https://getcaddy.com | bash -s personal http.cors
+
+1. Copy the PR2 description in a place where it can be served from:
+
+        mkdir ~/pr2_meshes -p
+        cp -r /opt/ros/kinetic/share/pr2_description ~/pr2_meshes
+
+1. Create a `Caddyfile` in that directory with the following contents:
+
+        cd ~/pr2_meshes
+        vi Caddyfile
+
+    Caddyfile:
+
+
+        0.0.0.0:8001
+        cors
+
+1. Start Caddy to serve the meshes on port 8081:
+
+        caddy
+
+1. Start a Gazebo simulation of a PR2 robot, this will open a Gazebo window with
+   a PR2 in an empty world.
+
+        roslaunch pr2_gazebo pr2_empty_world.launch
+
+1. Start the ROSBridge websocket server, which defaults to port 9090:
+
+        roslaunch rosbridge_server rosbridge_websocket.launch
+
+1. Start a TF2 republisher for the web:
+
+        rosrun tf2_web_republisher tf2_web_republisher
+
+1. Now refresh the browser. This time no error should be printed to the console.
+
+1. Click on the `+` and add a "Robot model", the PR2 should show up in the scene.
```

Comment:
I still can't see the robot in Firefox after following these steps; Caddy, Gazebo, rosbridge and tf2_web_republisher are running. I don't see any errors in the console either. Any ideas?

Reply:
Humm are you running this branch? I added a new dependency.

Do all the URLs under the "Global options" menu seem correct?

Reply:
Yes, the grid display works, and I'm in this branch.
I just went through all the steps again but it still doesn't work for me. The menu options seem correct:
```
{
  "globalOptions": {
    "background": "#111111",
    "colladaLoader": "collada2",
    "colladaServer": "http://localhost:8001/",
    "fixedFrame": "/base_link",
    "url": "ws://localhost:9090",
    "videoServer": "http://localhost:9999"
  },
  "sidebarOpened": true,
  "displays": [
    {
      "isShown": true,
      "name": "Grid",
      "options": {
        "cellSize": "1",
        "color": "#cccccc",
        "numCells": "10"
      },
      "type": "grid"
    },
    {
      "isShown": true,
      "name": "Robot model",
      "options": {
        "param": "robot_description"
      },
      "type": "urdf"
    }
  ]
}
```

My guess is that there's some issue when serving the meshes. I'll take another look and let you know.

Reply:
Well, I started from scratch and ran the steps in a fresh docker container but I got the same results. I reviewed the documentation ([ros-rviz wiki](https://github.com/jstnhuang/ros-rviz/wiki/User-guide) and [hcrlab wiki](https://github.com/hcrlab/wiki/wiki/Resources%3A-Web-Development%3A-Serving-URDF) but I didn't find substantial differences with the proposed steps.

I don't see any errors in the console when I add the robot model (only some warnings related to a missing animation, but I don't think it's related).

@chapulina do you see anything if you try to open a new tab with the URL of the meshes you are serving? (i.e. `http://localhost:8001`) I get a `404` when I do it. I'm using Firefox Quantum 63.0; what browser are you using?

Reply:
And one more update: I also tried these instructions https://github.com/hcrlab/wiki/wiki/Resources%3A-Web-Development%3A-Serving-URDF#in-production. When I go to `http://localhost:8001` in the browser I can now see the directory with the files, so it seems the web server is working; nevertheless, I still don't see the robot in the screen.

Perhaps there's an issue with my browser version?

Reply:
I'm trying the setup on a new computer to double-check things. Are you able to get the meshes doing something like this:

`wget http://0.0.0.0:8001/pr2_description/meshes/torso_v0/torso.stl`

Reply:
Some other thoughts:

* Since the robot is showing up really dark with Loader 2, try changing the background color to something other than black, maybe the robot is there but you can't see it
* Does the robot show up on RViz when you add `RobotModel`?
* Can you upload the logs of all your terminals somewhere and add a link here? Maybe there's some sneaky error message somewhere

Reply:
> Are you able to get the meshes doing something like this:

Yes, I can get it and I can also browse through the files in Firefox if I go to `http://localhost:8001`

> Since the robot is showing up really dark with Loader 2, try changing the background color to something other than black, maybe the robot is there but you can't see it

I tried both loaders and I also changed the background to white just in case, but I still don't see it.

> Does the robot show up on RViz when you add RobotModel?

Yes, it does! However, when I add the model I see a bunch of errors like these:
```
TIFFFetchNormalTag: Warning, Incompatible type for "RichTIFFIPTC"; tag ignored.
TIFFFieldWithTag: Internal error, unknown tag 0x829a.
```

And here's the logs:
[rviz_terminal.log](https://github.com/osrf/polymer-ros-rviz/files/2553258/rviz_terminal.log)
[rosbridge_terminal_.log](https://github.com/osrf/polymer-ros-rviz/files/2553259/rosbridge_terminal_.log)
[gazebo_terminal.log](https://github.com/osrf/polymer-ros-rviz/files/2553260/gazebo_terminal.log)

`tf2_web_republisher` has no logs at all. The only "suspicious" things I see are those `TIFF` tag logs, both in RViz and Gazebo terminals.

Reply:
Thanks for the logs, I can't spot anything wrong in them besides the tiff warnings, which I do not get, but I suspect they're unrelated.

One more thing you could check. On your browser's developer tools' network tab, the moment you choose the `RobotModel` option on the menu, do you see a bunch of GET requests for mesh files like this:

![polymer_get_meshes](https://user-images.githubusercontent.com/5751272/48085751-02aa2980-e1b0-11e8-9f5d-9e0bb427f514.png)

Reply:
Well, that doesn't happen on my side.

I can browse the files in a separate tab and I can get them with `wget`, so I guess they are being served properly:
![image](https://user-images.githubusercontent.com/2480899/48088522-04d8ac00-e1e1-11e8-8d28-d774033a2368.png)

I tried with `localhost`, my computer's IP address and `0.0.0.0`; all of these work in the separate tab, but not on web-rviz's collada server URL.

Reply:
Cool, so it looks like the browser is never asking for the model. I think one place to begin to debug would be to check if [_updateDisplay](https://github.com/osrf/polymer-ros-rviz/blob/master/ros-rviz-urdf.html#L61) is being called when you click on "Robot model". Depending on whether it is or not, the problem may be in the code in this repo, or on upstream code being used.

Reply:
Good catch.
The callback in `getParam.get` https://github.com/osrf/polymer-ros-rviz/blob/203a0d67e4fbf04b968e7fead9eb4a6be626e568/ros-rviz-urdf.html#L72 is never called.

The parameter is `robot_description`, and it exists if I get in in the console. So basically there's a problem retrieving it through roslib. I'll dig into it a bit.